aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -1162,6 +1162,60 @@ Tensor take_backward(const Tensor& grad, const Tensor& input, const Tensor& inde
   return at::zeros_like(input).put_(index, grad, true);
 }
 
+namespace {
+
+inline std::tuple<Tensor, Tensor, int64_t> _take_along_dim_helper(
+    const Tensor& self,
+    const Tensor& indices,
+    int64_t dim) {
+  TORCH_CHECK(
+      self.dim() == indices.dim(),
+      "self and indices should have same number of dimensions")
```

Comment:
"torch.take_along_dim(): input and indices should have the same number of dimensions, but got X dimensions for input, and Y dimensions for indices"

Remember to start with the name of the function and that we use "input" instead of "self" in our user-facing documentation (which is confusing, I know).

Reply:
Also check that the tensors are on the same device. Is there any dtype validation that should occur, too?

We should get in the habit of thinking: device, dtype, shape checking when implementing operators. (I sometimes forget, too!)

Reply:
For `indices`, we verify the `dtype` to be `Long`.

For `device`, I'll add a test to confirm. 

Thanks!

Reply:
Adding the device check upfront is OK because it will provide a clearer error message. While we could rely on gather's device checking, if the device check is tripped in gather() then the error message should say "torch.gather(): ...", which would be more confusing to users since it's not the function they called.

Reply:
I'll add a check upfront.

However at the moment,
for `torch.gather`,  check for all tensors being on same device is carried out by TensorIterator and it prints the message without the function name. (Maybe in the redesign of TensorIterator it should also take the API name to print it in warnings and errors).